File: thrift/ruby/lib/server/progress_monitor.rb
#
#   Copyright (c) 2012 VMware, Inc. All Rights Reserved.
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0

#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#

require 'thrift'
require 'chef'
require 'chef/knife'
require 'ironfan'
require 'ironfan/monitor'
require 'simple_logger'

module Software
  module Mgmt
    module Thrift
      module ProgressMonitor
        include Ironfan::Monitor

        ERROR_BOOTSTRAP_FAILURE = "Bootstrapping cluster %s failed"

        def log
          @log ||= SimpleLogger.new_logger(STDOUT)
        end

        def getClusterOperationStatus(targetName, withDetail = false)
          if(withDetail)
            clusterDetail = OperationStatusWithDetail.new
          end

          clusterName = fetchClusterName(targetName)
          clusterMain = OperationStatus.new
          clusterMain.total = 0
          clusterMain.success = 0
          clusterMain.failure = 0
          clusterMain.running = 0
          clusterMain.finished = false
          clusterMain.succeed = false
          clusterMain.progress = 0
          clusterMain.error_msg = ''
          if(withDetail)
            clusterData = ClusterData.new
            clusterData.clusterName = clusterName
            groups = Hash.new

            nodes = fetchClusterNodes(targetName)
            nodes.each do |node|
              server = node['provision'] || {}
              # create groups
              facet_name = node['facet_name']
              group = groups[facet_name] || GroupData.new
              group.groupName ||= facet_name
              group.instances ||= []
              serverData = convert2Data(server)
              group.instances << serverData
              groups[facet_name] = group
              # create cluster summary
              clusterMain.success += 1 if serverData.finished and serverData.succeed
              clusterMain.failure += 1 if serverData.finished and !serverData.succeed
              clusterMain.running += 1 if !serverData.finished
              clusterMain.progress += serverData.progress || 0
            end
            clusterMain.total = nodes.length
            clusterMain.progress /= clusterMain.total if clusterMain.total != 0
            clusterMain.finished = (clusterMain.running == 0)
            clusterMain.succeed = (clusterMain.success == clusterMain.total)
            clusterMain.error_msg = ERROR_BOOTSTRAP_FAILURE % clusterName if clusterMain.finished and !clusterMain.succeed

            clusterData.groups = groups
            clusterDetail.clusterData = clusterData
            clusterDetail.operationStatus = clusterMain
          end
          cluster = clusterMain
          if(withDetail)
            cluster = clusterDetail
          end
          cluster
        end

        def convert2Data(server)
          serverData = ServerData.new
          serverData.status = server['status']
          serverData.action = server['action']
          serverData.bootstrapped = server['bootstrapped']
          serverData.created = server['created']
          serverData.deleted = server['deleted']
          serverData.error_code = server['error_code']
          serverData.error_msg = server['error_msg']
          serverData.finished = server['finished']
          serverData.ha = server['ha']
          serverData.hostName = server['hostname']
          serverData.ipAddress = server['ip_address']
          serverData.name = server['name']
          serverData.physicalHost = server['physical_host']
          serverData.progress = server['progress']
          serverData.rack = server['rack']
          serverData.succeed = server['succeed']
          serverData
        end

        def fetchClusterName(targetName)
          return targetName.split('-')[0]
        end

        def fetchClusterNodes(targetName)
          nodes = []
          clusterName = fetchClusterName(targetName)
          # This method will be called about every 6 seconds and it will retrieve all Chef Nodes of this cluster.
          # So use Chef Partial Search API to get the needed Chef Node attributes only. This can reduce server response time.
          # See http://www.opscode.com/blog/2012/10/31/introducing-partial-search-for-opscode-hosted-chef/
          query = Chef::Search::Query.new
          query.partial_search(:node, "cluster_name:#{clusterName}", keys: { name: ['name'], facet_name: ['facet_name'], provision: ['provision'] }) do |n|
            nodes.push(n) if n['name'].start_with?(targetName)
          end
          nodes.sort_by! { |n| n['name'] }
        end
      end
    end
  end
end

File: server/cluster-mgmt/src/main/java/com/vmware/bdd/service/job/software/ProgressMonitor.java
/***************************************************************************
 * Copyright (c) 2012-2013 VMware, Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
package com.vmware.bdd.service.job.software;

import org.apache.log4j.Logger;

import com.vmware.bdd.manager.ClusterEntityManager;
import com.vmware.bdd.service.job.StatusUpdater;
import com.vmware.bdd.software.mgmt.impl.SoftwareManagementClient;
import com.vmware.bdd.software.mgmt.thrift.OperationStatusWithDetail;
import com.vmware.bdd.utils.TracedRunnable;

/**
 * @author Jarred Li
 * @version 0.8
 * @since 0.8
 * 
 */
public class ProgressMonitor extends TracedRunnable {
   private static final Logger logger = Logger.getLogger(ProgressMonitor.class);
   private static final long QUERY_INTERVAL_DEFAULT = 1000 * 10; // 10 seconds
   private static final long QUERY_INTERVAL_LONG = 1000 * 60 * 5; // 5 minutes
   private static final long QUERY_INTERVAL_MAX = 1000 * 60 * 15; // 15 minutes
   private static final int BIG_CLUSTER_NODES_COUNT = 100;

   private String targetName;
   private StatusUpdater statusUpdater;
   private String lastErrorMsg = null;
   private long queryInterval = QUERY_INTERVAL_DEFAULT;
   private ClusterEntityManager clusterEntityMgr;
   private volatile boolean stop;

   public ProgressMonitor(String targetName, StatusUpdater statusUpdater,
         ClusterEntityManager clusterEntityMgr) {
      this.targetName = targetName;
      this.statusUpdater = statusUpdater;
      this.clusterEntityMgr = clusterEntityMgr;
   }

   /**
    * @return should stop monitoring?
    */
   public boolean isStop() {
      return stop;
   }

   /**
    * @param stop
    *           stop monitoring
    */
   public void setStop(boolean stop) {
      this.stop = stop;
   }

   public String getLastErrorMsg() {
      return lastErrorMsg;
   }

   private void setLastErrorMsg(String errorMsg) {
      this.lastErrorMsg = errorMsg;
   }

   /* (non-Javadoc)
    * @see com.vmware.bdd.command.TracedRunnable#doWork()
    */
   @Override
   public void doWork() throws Exception {
      boolean exit = false;

      logger.info("start monitor operation progress for target " + targetName);
      OperationStatusWithDetail detailedStatus = null;
      SoftwareManagementClient monitorClient = new SoftwareManagementClient();
      monitorClient.init();

      while (!exit) {
         try {
            Thread.sleep(queryInterval);
         } catch (InterruptedException e) {
            logger.info("monitor thread is waked up by caller to stop monitoring");
            stop = true;
         }
         if (stop) {
            logger.info("received stop signal, will do one more query then exit, so as to retrieve the final progress after Ironfan exits.");
            exit = true;
         }

         logger.info("progress query started");
         detailedStatus = monitorClient.getOperationStatusWithDetail(targetName);
         if (null == detailedStatus) {
            logger.error("Failed to query progress. Something wrong with the Monitor Service?");
            break;
         }
         logger.info("progress finished? " + detailedStatus.getOperationStatus().isFinished());
         logger.debug(detailedStatus.toString());
         logger.info("progress query completed");
         if (detailedStatus.getOperationStatus().getProgress() < 100) {
            int progress = detailedStatus.getOperationStatus().getProgress();
            statusUpdater.setProgress(((double) progress) / 100);
         }
         setLastErrorMsg(detailedStatus.getOperationStatus().getErrorMsg());
         clusterEntityMgr.handleOperationStatus(targetName.split("-")[0], detailedStatus);

         // for large scale cluster (100+ nodes), don't need to query too frequent which will cause too much overhead on Chef Server
         if (queryInterval == QUERY_INTERVAL_DEFAULT) {
            int size = detailedStatus.getClusterData().getClusterSize();
            if (size > BIG_CLUSTER_NODES_COUNT) {
               queryInterval = Math.min(QUERY_INTERVAL_MAX, QUERY_INTERVAL_LONG * (size / BIG_CLUSTER_NODES_COUNT));
               logger.info("progress query interval set to " + queryInterval / 1000 + "s because this cluster has " + size + " nodes");
            }
         }
      }
      if (monitorClient != null) {
         monitorClient.close();
      }
   }

   /* (non-Javadoc)
    * @see com.vmware.bdd.command.TracedRunnable#onStart()
    */
   @Override
   public void onStart() throws Exception {
      // TODO Auto-generated method stub

   }

   /* (non-Javadoc)
    * @see com.vmware.bdd.command.TracedRunnable#onFinish()
    */
   @Override
   public void onFinish() throws Exception {
      // TODO Auto-generated method stub

   }

   /* (non-Javadoc)
    * @see com.vmware.bdd.command.TracedRunnable#onException(java.lang.Throwable)
    */
   @Override
   public void onException(Throwable t) {
      logger.error(t.getMessage());
   }

}
